### PR TITLE
C/C++: Add the latest 4 versions of Boost 1.66.0 - 1.69.0 (Linux, Windows)

### PR DESCRIPTION
**Pull request by gekeugen**
**Issue:** [C/C++: Add the latest 4 versions of Boost 1.66.0 - 1.69.0 (Linux, Windows) #263](https://github.com/Microsoft/azure-pipelines-image-generation/issues/263)

This PR contains changes to preinstall Boost on Linux and Windows VM (VS2017 Server 2016 and VS2019 Server 2019 images) with the following compilation options:
* debug and release
* x32/x64
* multithreaded
* link static and shared
* runtime link static and shared

- Boost `1.66.0`, `1.67.0`, `1.68.0`, `1.69.0` will be added to the VM image
- Boost build artifacts are located in tool cache. This PR adds logic to unpack them from the tool cache.
- Adding 4 Boost versions will increase VM image size by 18GB (Windows image) and 8GB (Ubuntu image)
- Adding 4 Boost versions will increase VM image generation time by 25mins (Windows image) and 13mins (Ubuntu image): it is the time to download tool cache archives and unpack them.
- The following environment variables will be available for customers:
    - `BOOST_ROOT_1_66_0`
    - `BOOST_ROOT_1_67_0`
    - `BOOST_ROOT_1_68_0`
    - `BOOST_ROOT_1_69_0`
    - `BOOST_ROOT` (points to `BOOST_ROOT_1_69_0`)

### CODE REVIEW

File: images/linux/scripts/installers/boost.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        echo "BOOST_ROOT=$BOOST_LIB/$BOOST_VERSION" | tee -a /etc/environment
+    fi
+
+    DocumentInstalledItem "Boost C++ Libraries $BOOST_VERSION"
+done
+
+# Deleting archives with Boost Libraries
+rm -rf $BOOST_ZIP_PATH
+
+# Document what was added to the image
```
Comment (sgrebnov) [inline images/linux/scripts/installers/boost.sh]: Should this be removed as we do documentation step before this - looks inconsistent? /cc @max-lobanov

File: images/win/scripts/Installers/Install-Boost.ps1 [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+$ToolCache = 'C:\hostedtoolcache\windows\boost'
+$BoostDirectory = Join-Path -Path $env:ProgramFiles -ChildPath "Boost"
+$BoostVersionsToInstall = $env:BOOST_VERSIONS.split(',')
+$BoostDefault = $env:BOOST_DEFAULT
+
+foreach($BoostVersion in $BoostVersionsToInstall)
+{
+    $ZipName = Join-Path -Path $ToolCache -ChildPath "boost_${BoostVersion}_msvc-14.1.zip"
+
+    if (Test-Path $ZipName)
```
Comment (sgrebnov) [inline images/win/scripts/Installers/Install-Boost.ps1]: Nit: better to do like below. Avoid unnecessary if/else blocks and corresponding indents

```
if(!Test-Path $ZipName ) 
{
   Write-Host "$ZipName not found"
   exit(1)
}

Write-Host "Expanding $ZipName"
```

File: images/win/scripts/Installers/Install-Boost.ps1 [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        Write-Host "Expanding $ZipName"
+
+        $BoostInstallationDir = Join-Path -Path $BoostDirectory -ChildPath $BoostVersion
+        # Expand-Archive slower for 70% than 7z
```
Comment (sgrebnov) [inline images/win/scripts/Installers/Install-Boost.ps1]: Very good comment and approach - like it 👍

File: images/linux/scripts/installers/boost.sh (hunk 1 above)
Comment (max-lobanov) [reply images/linux/scripts/installers/boost.sh]: It makes sense +1

**Comment by mloskot** [comment]
The docs still list only Boost 1.69

https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2017-Server2016-Readme.md
https://github.com/Microsoft/azure-pipelines-image-generation/blob/master/images/win/Vs2019-Server2019-Readme.md

Please, could this list be kept up to date?
The 'silent' changes to AzP images do have potential to break builds  e.g.
https://developercommunity.visualstudio.com/content/problem/525582/ubuntu-1604-image-changed-and-comes-with-usrlocals.html

**Comment by maxim-lobanov** [comment]
@mloskot , could you please create the new issue about it?

**Comment by mloskot** [comment]
@maxim-lobanov Where?
In fact, I have
- https://developercommunity.visualstudio.com/idea/556787/stop-setting-boost-root-as-predefined-variable.html
- https://developercommunity.visualstudio.com/comments/527647/view.html
and others have:
- https://developercommunity.visualstudio.com/content/problem/495493/setting-boost-root-in-the-vs2017-hosted-agent-brok.html

but nobody seems to listen to the annoyed ones.

**Pull request merged**